README
= SDL (Simple Declarative Language)

SDL version supported: 1.3

Site:: http://www.ikayzo.org/confluence/display/SDL/Home
Downloads:: http://www.ikayzo.org/confluence/display/SDL/Downloads

Users mailing list:: sdl-users@ikayzo.org
Developers mailing list:: sdl-developers@ikayzo.org


== Getting Started with SDL4R

To get the Ruby Gem:

    gem install sdl4r

Then, you can start reading SDL documents:

    require 'pathname'
    require 'sdl4r'

    root = SDL4R::read(Pathname.new("my_directory/my_config.sdl"))
    puts root.attribute("port")

Or you can create SDL documents with the API:

    require 'fileutils'
    require 'sdl4r'

    root = SDL4R::Tag.new("root") do
        new_child("server") do
            set_attribute("port", 1234)
        end
    end
    File.open("my_directory/my_config.sdl", "w") { |io|
        io.write(root.children_to_string)
    }

which will write the following in your file:

    server port=1234

== SDL Documents

SDL documents are made up of Tags. A Tag contains

* a name (if not present, the name "content" is used)
* a namespace (optional)
* 0 or more values (optional)
* 0 or more attributes (optional)
* 0 or more children (optional)

For the SDL code:

  size 4
  smoker false


Assuming this code is in a file called <tt>values.sdl</tt>, the values can be read
using the following code (ignoring exceptions):

  root = Tag.new("root").read(Pathname.new("values.sdl"))
  size = root.child("size").value
  smoker = root.child("smoker").value

A tag is basically a data structure with a list of values, a map of
attributes, and (if it has a body) child tags.  In the example above, the
<tt>values.sdl</tt> file is read into a tag called "root".  It has two children
(tags) called "size" and "smoker".  Both these children have one value, no
attributes, and no bodies.

SDL is often used for simple key-value mappings.  To simplify things Tag
has the methods getValue and setValue which operate on the first element in
the values list.  Also notice SDL understands types which are determined
using type inference.

The example above used the simple format common in property files:

  name value

The full SDL tag format is:

  namespace:name value_list attribute_list {
    children_tags
  }

where value_list is zero or more space separated SDL literals and
attribute_list is zero or more space separated <tt>(namespace:)key=value</tt> pairs.
The name, namespace, and keys are SDL identifiers.  Values are SDL literals.
Namespace is optional for both tag names and attributes.  Tag bodies are also
optional.  SDL identifiers begin with a unicode letter or an underscore (_)
followed by zero or more unicode letters, numbers, underscores (_),
dashes (-) and periods (.).

Tags without bodies are terminated by a new line character (\n) and may be
continue onto the next line by placing a backslash (\) at the end of the
line.  Tags may be nested to an arbitrary depth.  SDL ignores all other white
space characters between tokens.  Although nested blocks are indented by
convention, tabs have no significance in the language.

== Anonymous Tags

SDL also supports anonymous tags which are assigned the name "content".
An anonymous tag starts with a literal and is followed by zero or more
additional literals and zero or more attributes.  The examples section below
demonstrates the use of anonymous tags.

    greetings {
       "hello" language="English"
    }

    # If we have a handle on the "greetings" tag we can access the
    # anonymous child tag by calling
    #    Tag child1 = greetingTag.getChild("content");

== String literals

There are two ways to write String literals.

=== Starting and ending with double quotes (")

Double quotes, backslash characters (\), and new lines (\n) within this type of String literal
must be escaped like so:

  file "C:\\folder\\file.txt"
  say "I said \"something\""

This type of String literal can be continued on the next line by placing a
backslash (\) at the end of the line like so:

  line "this is a \
    long string of text"

White space before the first character in the second line will be ignored.

=== Starting and ending with a backquote (`)

This type of string literal
can only be ended with a second backquote (`).  It is not necessary (or
possible) to escape any type of character within a backquote string
literal. This type of literal can also span lines. All white spaces are
preserved including new lines.

Examples:

  file `C:\folder\file.txt`
  say `I said "something"`
  regex `\w+\.suite\(\)`
  long_line `This is
      a long line
      fee fi fo fum`

Note: SDL interprets new lines in `` String literals as a single new line
character (\n) regarless of the platform.

== Binary literals

Binary literals use base64 characters enclosed in square brackets ([]).
The binary literal type can also span lines.  White space is ignored.

Examples:
  key [sdf789GSfsb2+3324sf2] name="my key"
  image [
    R3df789GSfsb2edfSFSDF
    uikuikk2349GSfsb2edfS
    vFSDFR3df789GSfsb2edf
  ]
  upload from="ikayzo.com" data=[
    R3df789GSfsb2edfSFSDF
    uikuikk2349GSfsb2edfS
    vFSDFR3df789GSfsb2edf
  ]

== Date and Time Literals

SDL supports date, time span, and date/time literals.  Date and Date/Time
literals use a 24 hour clock (0-23).  If a timezone is not specified, the
default locale's timezone will be used.

Examples:

* create a tag called "date" with a date value of Dec 5, 2005
    date 2005/12/05

* various time span literals
    hours 03:00:00
    minutes 00:12:00
    seconds 00:00:42
    short_time 00:12:32.423 # 12 minutes, 32 seconds, 423 milliseconds
    long_time 30d:15:23:04.023 # 30 days, 15 hours, 23 mins, 4 secs, 23 millis
    before -00:02:30 # 2 hours and 30 minutes ago

* a date time literal
    in_japan 2005/12/05 14:12:23.345-JST

== Literal Types

SDL 1.0 has thirteen literal types (parenthesis indicate optional components)

1. string (unicode) - examples: <tt>"hello"</tt> or <tt>`aloha`</tt>
2. character (unicode) - example: <tt>'/'</tt>
   Note: \uXXXX style unicode escapes are not supported (or needed because sdl files are UTF8)
3. integer (32 bits signed) - example: <tt>123</tt>
4. long integer (64 bits signed) - examples: <tt>123L</tt> or <tt>123l</tt>
5. float (32 bits signed) - examples <tt>123.43F</tt> <tt>123.43f</tt>
6. double float (64 bits signed) - example: <tt>123.43</tt> or <tt>123.43d</tt> or <tt>123.43D</tt>
7. decimal (128+ bits signed) - example: <tt>123.44BD</tt> or <tt>123.44bd</tt>
8. boolean - examples: <tt>true</tt> or <tt>false</tt> or <tt>on</tt> or <tt>off</tt>
9. date yyyy/mm/dd - example <tt>2005/12/05</tt>
10. date time yyyy/mm/dd hh:mm(:ss)(.xxx)(-ZONE)
    example - <tt>2005/12/05 05:21:23.532-JST</tt>
    notes: uses a 24 hour clock (0-23), only hours and minutes are mandatory
11. time span using the format (d:)hh:mm:ss(.xxx)
    notes: if the day component is included it must be suffixed with a lower case 'd'
    examples
        12:14:42 # (12 hours, 14 minutes, 42 seconds)
        00:09:12 # (9 minutes, 12 seconds)
        00:00:01.023 # (1 second, 23 milliseconds)
        23d:05:21:23.532 # (23 days, 5 hours, 21 minutes, 23 seconds, 532 milliseconds)
12. binary [base64] example - <tt>[sdf789GSfsb2+3324sf2]</tt>
13. <tt>null</tt>


Timezones must be specified using a valid time zone ID (ex. America/Los_Angeles), three letter
abbreviation (ex. HST), or GMT(+/-)hh(:mm) formatted custom timezone (ex. GMT+02 or GMT+02:30)

These types are designed to be portable across Java, .NET, and other popular platforms.

== SDL Comments

SDL supports four comment types.

  1. // single line comments identicle to those used in Java, C, etc. // style
    comments can occur anywhere in a line.  All text after // up to the new line
    will be ignored.
  2. # property style comments.  They work the same way as //
  3. -- separator comments useful for visually dividing content.  They work the same way as //
  4. Slash star (/*) style multiline comments.  These begin with a slash
    star and end with a star slash.  Everything in between is ignored.


== Example

An example SDL file:

    # a tag having only a name
    my_tag

    # three tags acting as name value pairs
    first_name "Akiko"
    last_name "Johnson"
    height 68

    # a tag with a value list
    person "Akiko" "Johnson" 68

    # a tag with attributes
    person first_name="Akiko" last_name="Johnson" height=68

    # a tag with values and attributes
    person "Akiko" "Johnson" height=60

    # a tag with attributes using namespaces
    person name:first-name="Akiko" name:last-name="Johnson"

    # a tag with values, attributes, namespaces, and children
    my_namespace:person "Akiko" "Johnson" dimensions:height=68 {
        son "Nouhiro" "Johnson"
        daughter "Sabrina" "Johnson" location="Italy" {
            hobbies "swimming" "surfing"
            languages "English" "Italian"
            smoker false
        }
    }

    ------------------------------------------------------------------
    // (notice the separator style comment above...)

    # a log entry
    #     note - this tag has two values (date_time and string) and an
    #            attribute (error)
    entry 2005/11/23 10:14:23.253-GMT "Something bad happened" error=true

    # a long line
    mylist "something" "another" true "shoe" 2002/12/13 "rock" \
        "morestuff" "sink" "penny" 12:15:23.425

    # a long string
    text "this is a long rambling line of text with a continuation \
       and it keeps going and going..."

    # anonymous tag examples

    files {
        "/folder1/file.txt"
        "/file2.txt"
    }

    # To retrieve the files as a list of strings
    #
    #     List files = tag.getChild("files").getChildrenValues("content");
    #
    # We us the name "content" because the files tag has two children, each of
    # which are anonymous tags (values with no name.)  These tags are assigned
    # the name "content"

    matrix {
        1 2 3
        4 5 6
    }

    # To retrieve the values from the matrix (as a list of lists)
    #
    #     List rows = tag.getChild("matrix").getChildrenValues("content");


Example of getting the "location" attribute from the "daughter" tag
above (ignoring exceptions)

     root = SDL4R.read(Pathname.new("myfile.sdl"))
     daughter = root.child("daughter", true) // recursive search
     location = daughter.attribute("location")

SDL is normally stored in a file with the .sdl extension.  These files
should always be encoded using UTF8.  SDL fully supports unicode in
identifiers and literals.

== Ruby and SDL types

The following list gives what types are used in Ruby in order to represent SDL types.

*SDL*:: *Ruby*
unicode string:: String
unicode character:: single-character String
integer (32 bits signed):: Integer (Fixnum or Bignum)
long integer (64 bits signed):: Integer (Fixnum or Bignum)
float (32 bits signed):: Float
double float (64 bits signed):: Float
decimal (128+ bits signed):: BigDecimal
boolean:: true (TrueClass) and false (FalseClass)
date (day):: Date

CHANGELOG
== v0.9.8 (建設中)

== v0.9.7 (26-auf-2010)

=== Major changes:

* It could prove a hassle to use the block idiom of Tag.new or Tag.new_child when handling another
  set of data. Therefore, if the given block takes a (not-optional) parameter, the context remains
  the context of the calling code.

    class Elephant

      def initialize
         @weight = 1000
      end

      def to_sdl
        SDL4R::Tag.new("elephant") do |tag|
          tag.set_attribute("weight", @weight)
          tag << "I love pink ones."
        end
      end

    end

    puts Elephant.new.to_sdl.to_s # ==> elephant "I love pink ones." weight=1000

* Tag.to_xml_string() has been refactor to take an option Hash (backward incompatible):

  [:line_prefix] a text prefixing each line (default: "")
  [:uri_by_namespace] a Hash giving the URIs for the namespaces
  [:indent] text specifying one indentation (default: "\t")
  [:eol] end of line expression (default: "\n")
  [:omit_null_attributes]
    if true, null/nil attributes are not exported (default: false). Otherwise, they are exported
    as follows:
      tag attr="null"

* Fixed the parsing of line continuations with anti-slashes.
* Fixed the parsing of double-quoted strings. It is better at detecting badly formed strings, now.
* Fixed the reporting of lines and columns in errors.
  Although the column might be slightly off sometimes, the line should be correct.


=== Minor changes:

* The RDoc is now generated with the template Hanna[http://github.com/mislav/hanna/tree/master].
* Tag.<< now behaves with Enumerables (except Strings) as it does with Arrays.

== v0.9.6 (10-aug-2010)

=== Major changes:

* Tag.set_attribute(): fixed as it failed if the value was nil.
* Tag.write(): now returns the given output (especially necessary if a String was given).
* Tag.write(): fixed as it failed when writing children without including the root Tag.
* Tag.write(): it is now possible to give a StringIO as an output

== v0.9.5 (8-aug-2010)

=== Major changes:

* Fixes in order to make SDL4R work in Ruby 1.9 (mainly concerning encodings and dates).
* Commented out debug code (writing to files) inside sdl4r_test.rb.

=== Minor changes:

* General & contact info in the README

== v0.9.4 (6-aug-2010)

=== Major changes:

* The use of BigDecimal (standard) has been preferred to Flt::DecNum.
* "sdl4r/sdl.rb" has been renamed to "sdl4r/sdl4r.rb"
* A sdl4r.rb has been added so that users can actually write (as the documentation pretended):

    require 'sdl4r'

* SDL4R#coerce_or_fail has been implemented, which means that random objects cannot be added to a
  Tag as before. Rational instances are coerced using Rational#to_f.
* Added usual time methods to SdlTimeSpan: day(), hour(), min(), sec(), usec()

== v0.9.3

=== Major changes

* Tag: any method that use to take namespace/name parameters now follows the natural order
  "namespace" and then "name":

    Tag.new "ns1", "tag1"

  WARNING: this was the contrary before!!!
  Also, the "recursive" parameter should always be at the first place now.

* It is now possible to override SDL4R::new_date_time in order to create Time instances when parsing
  rather than DateTime instances.

lib/sdl4r/parser.rb
#!/usr/bin/env ruby -w
# encoding: UTF-8

#--
# Simple Declarative Language (SDL) for Ruby
# Copyright 2005 Ikayzo, inc.
#
# This program is free software. You can distribute or modify it under the
# terms of the GNU Lesser General Public License version 2.1 as published by
# the Free Software Foundation.
#
# This program is distributed AS IS and WITHOUT WARRANTY. OF ANY KIND,
# INCLUDING MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, contact the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#++


module SDL4R

  require 'base64'

  require File.dirname(__FILE__) + '/sdl_binary'
  require File.dirname(__FILE__) + '/sdl_time_span'
  require File.dirname(__FILE__) + '/sdl_parse_error'
  require File.dirname(__FILE__) + '/parser/tokenizer'

  # The SDL parser.
  #
  # In Ruby 1.8, in order to enable UTF-8 support, you may have to declare the following lines:
  #
  #   $KCODE = 'u'
  #   require 'jcode'
  #
  # This will give you correct input and output and correct UTF-8 "general" sorting.
  # Alternatively you can use the following options when launching the Ruby interpreter:
  #
  #   /path/to/ruby -Ku -rjcode
  #
  # == Authors
  # Daniel Leuck, Philippe Vosges
  #
  class Parser
			
    # Passed to parse_error() in order to specify an error that occured on no specific position
    # (column).
    UNKNOWN_POSITION = -2
	 	
    # Creates an SDL parser on the specified +IO+.
    #
    #   IO.open("path/to/sdl_file") { |io|
    #     parser = SDL4R::Parser.new(io)
    #     tags = parser.parse()
    #   }
    #
    def initialize(io)
      raise ArgumentError, "io == nil" if io.nil?
			
      @tokenizer = Tokenizer.new(io)
    end
		
    # Parses the underlying +IO+ and returns an +Array+ of +Tag+.
    # 
    # ==Errors
    # [IOError] If a problem is encountered with the IO
    # [SdlParseError] If the document is malformed
    def parse
      tags = []
			
      while tokens = @tokenizer.read_line_tokens()
        if tokens.last.type == :START_BLOCK
          # tag with a block
          tag = construct_tag(tokens[0...-1])
          add_children(tag)
          tags << tag

        elsif tokens.first.type == :END_BLOCK
          # we found an block end token that should have been consumed by
          # add_children() normally
          parse_error(
            "No opening block ({) for close block (}).",
            tokens.first.line,
            tokens.first.position)
        else
          # tag without block
          tags << construct_tag(tokens)
        end
      end
			
      @tokenizer.close()
			
      return tags
    end

    # Creates and returns the object representing a datetime (DateTime in the default
    # implementation). Can be overriden.
    #
    #   def new_date_time(year, month, day, hour, min, sec, time_zone_offset)
    #     Time.utc(year, month, day, hour, min, sec)
    #   end
    #
    def new_date_time(year, month, day, hour, min, sec, time_zone_offset)
      SDL4R::new_date_time(year, month, day, hour, min, sec, time_zone_offset)
    end
		
    private
		
    # Parses the children tags of +parent+ until an end of block is found.
    def add_children(parent)
      while tokens = @tokenizer.read_line_tokens()
        if tokens.first.type == :END_BLOCK
          return
					
        elsif tokens.last.type == :START_BLOCK
          # found a child with a block
          tag = construct_tag(tokens[0...-1]);
          add_children(tag)
          parent.add_child(tag)
					
        else
          parent.add_child(construct_tag(tokens))
        end
      end
			
      parse_error("No close block (}).", @tokenizer.line_no, UNKNOWN_POSITION)
    end
		
    # Construct a Tag (but not its children) from a string of tokens
    # 
    # Throws SdlParseError if some bad syntax is found.
    def construct_tag(tokens)
      raise ArgumentError, "tokens == nil" if tokens.nil?
      if tokens.empty?
        parse_error("Internal Error: empty token list", @tokenizer.line_no, UNKNOWN_POSITION)
      end
			
      first_token = tokens.first
      if first_token.literal?
        first_token = Token.new("content")
        tokens.insert(0, first_token)
				
      elsif first_token.type != :IDENTIFIER
        expecting_but_got(
          "IDENTIFIER",
          "#{first_token.type} (#{first_token.text})",
          first_token.line,
          first_token.position)
      end
			
      tag = nil
      if tokens.size == 1
        tag = Tag.new(first_token.text)
				
      else
        values_start_index = 1
        second_token = tokens[1]
				
        if second_token.type == :COLON
          if tokens.size == 2 or tokens[2].type != :IDENTIFIER
            parse_error(
              "Colon (:) encountered in unexpected location.",
              second_token.line,
              second_token.position)
          end
					
          third_token = tokens[2];
          tag = Tag.new(first_token.text, third_token.text)
          values_start_index = 3
					
        else
          tag = Tag.new(first_token.text)
        end
				
        # read values
        attribute_start_index = add_tag_values(tag, tokens, values_start_index)
				
        # read attributes
        if attribute_start_index < tokens.size
          add_tag_attributes(tag, tokens, attribute_start_index)
        end
      end
			
      return tag
    end
		
    #
    # @return The position at the end of the value list
    #
    def add_tag_values(tag, tokens, start)
      size = tokens.size()
      i = start;
			
      while i < size
        token = tokens[i]
				
        if token.literal?
          # if a DATE token is followed by a TIME token combine them
          next_token = ((i + 1) < size)? tokens[i + 1] : nil
          if token.type == :DATE && next_token && next_token.type == :TIME
            date = token.object_for_literal()
            time_zone_with_zone = next_token.object_for_literal()
						
            if time_zone_with_zone.day != 0
              # as there are days specified, it can't be a full precision date
              tag.add_value(date);
              tag.add_value(
                SdlTimeSpan.new(
                  time_zone_with_zone.day,
                  time_zone_with_zone.hour,
                  time_zone_with_zone.min,
                  time_zone_with_zone.sec))
							
							
              if time_zone_with_zone.time_zone_offset
                parse_error("TimeSpan cannot have a timeZone", t.line, t.position)
              end
							
            else
              tag.add_value(combine(date, time_zone_with_zone))
            end
						
            i += 1
						
          else
            value = token.object_for_literal()
            if value.is_a?(TimeSpanWithZone)
              # the literal looks like a time zone
              if value.time_zone_offset
                expecting_but_got(
                  "TIME SPAN",
                  "TIME (component of date/time)",
                  token.line,
                  token.position)
              end
							
              tag.add_value(
                SdlTimeSpan.new(
                  value.day,
                  value.hour,
                  value.min,
                  value.sec))
            else
              tag.add_value(value)
            end
          end
        elsif token.type == :IDENTIFIER
          break
        else
          expecting_but_got(
            "LITERAL or IDENTIFIER", token.type, token.line, token.position)
        end
				
        i += 1
      end
			
      return i
    end
		
    #
    # Add attributes to the given tag
    #
    def add_tag_attributes(tag, tokens, start)
      i = start
      size = tokens.size
			
      while i < size
        token = tokens[i]
        if token.type != :IDENTIFIER
          expecting_but_got("IDENTIFIER", token.type, token.line, token.position)
        end
        name_or_namespace = token.text;
				
        if i == (size - 1)
          expecting_but_got(
            "\":\" or \"=\" \"LITERAL\"",
            "END OF LINE.",
            token.line,
            token.position)
        end
				
        i += 1
        token = tokens[i]
        if token.type == :COLON
          if i == (size - 1)
            expecting_but_got(
              "IDENTIFIER", "END OF LINE", token.line, token.position)
          end
					
          i += 1
          token = tokens[i]
          if token.type != :IDENTIFIER
            expecting_but_got(
              "IDENTIFIER", token.type, token.line, token.position)
          end
          name = token.text
					
          if i == (size - 1)
            expecting_but_got("\"=\"", "END OF LINE", token.line, token.position)
          end
					
          i += 1
          token = tokens[i]
          if token.type != :EQUALS
            expecting_but_got("\"=\"", token.type, token.line, token.position)
          end
					
          if i == (size - 1)
            expecting_but_got("LITERAL", "END OF LINE", token.line, token.position)
          end
					
          i += 1
          token = tokens[i]
          if !token.literal?
            expecting_but_got("LITERAL", token.type, token.line, token.position)
          end
					
          if token.type == :DATE and (i + 1) < size and tokens[i + 1].type == :TIME
            date = token.get_object_for_literal()
            time_span_with_zone = tokens[i + 1].get_object_for_literal()
						
            if time_span_with_zone.days != 0
              expecting_but_got(
                "TIME (component of date/time) in attribute value",
                "TIME SPAN",
                token.line,
                token.position)
            else
              tag.set_attribute(name_or_namespace, name, combine(date, time_span_with_zone))
            end
						
            i += 1
          else
            value = token.object_for_literal();
            if value.is_a?(TimeSpanWithZone)
              time_span_with_zone = value
							
              if time_span_with_zone.time_zone_offset
                expecting_but_got(
                  "TIME SPAN",
                  "TIME (component of date/time)",
                  token.line,
                  token.position)
              end
							
              time_span = SdlTimeSpan.new(
                time_span_with_zone.day,
                time_span_with_zone.hour,
                time_span_with_zone.min,
                time_span_with_zone.sec)
							
              tag.set_attribute(name_or_namespace, name, time_span)
            else
              tag.set_attribute(name_or_namespace, name, value);
            end
          end
        elsif token.type == :EQUALS
          if i == (size - 1)
            expecting_but_got("LITERAL", "END OF LINE", token.line, token.position)
          end
					
          i += 1
          token = tokens[i]
          if !token.literal?
            expecting_but_got("LITERAL", token.type, token.line, token.position)
          end
					
          if token.type == :DATE and (i + 1) < size and tokens[i + 1].type == :TIME
            date = token.object_for_literal()
            time_span_with_zone = tokens[i + 1].object_for_literal()
						
            if time_span_with_zone.day != 0
              expecting_but_got(
                "TIME (component of date/time) in attribute value",
                "TIME SPAN",
                token.line,
                token.position)
            end
            tag.set_attribute(name_or_namespace, combine(date, time_span_with_zone))
						
            i += 1
          else
            value = token.object_for_literal()
            if value.is_a?(TimeSpanWithZone)
              time_span_with_zone = value
              if time_span_with_zone.time_zone_offset
                expecting_but_got(
                  "TIME SPAN",
                  "TIME (component of date/time)",
                  token.line,
                  token.position)
              end
							
              time_span = SdlTimeSpan.new(
                time_span_with_zone.day,
                time_span_with_zone.hour,
                time_span_with_zone.min,
                time_span_with_zone.sec)
              tag.set_attribute(name_or_namespace, time_span)
            else
              tag.set_attribute(name_or_namespace, value);
            end			
          end
        else
          expecting_but_got(
            "\":\" or \"=\"", token.type, token.line, token.position)
        end
				
        i += 1
      end
    end
		
    # Combines a simple Date with a TimeSpanWithZone to create a DateTime
    #
    def combine(date, time_span_with_zone)
      time_zone_offset = time_span_with_zone.time_zone_offset
      time_zone_offset = TimeSpanWithZone.default_time_zone_offset if time_zone_offset.nil?
			
      new_date_time(
        date.year,
        date.month,
        date.day,
        time_span_with_zone.hour,
        time_span_with_zone.min,
        time_span_with_zone.sec,
        time_zone_offset)
    end
		
    private
    ############################################################################
    ## Parsers for types
    ############################################################################
		
    def Parser.parse_string(literal)
      unless literal =~ /(^`.*`$)|(^\".*\"$)/m
        raise ArgumentError,
          "Malformed string <#{literal}>." +
          "	Strings must start and end with \" or `"
      end
			
      return literal[1..-2]
    end
		
    def Parser.parse_character(literal)
      unless literal =~ /(^'.*'$)/
        raise ArgumentError,
          "Malformed character <#{literal}>." +
          "	Character must start and end with single quotes"
      end
			
      return literal[1]
    end
		
    def Parser.parse_number(literal)
      # we use the fact that Kernel.Integer() and Kernel.Float() raise ArgumentErrors
      if literal =~ /(.*)(L)$/i
        return Integer($1)
      elsif literal =~ /([^BDF]*)(BD)$/i
        return BigDecimal($1)
      elsif literal =~ /([^BDF]*)(F|D)$/i
        return Float($1)
      elsif literal.count(".e") == 0
        return Integer(literal)
      else
        return Float(literal)
      end
    end
		
    # Parses the given literal into a returned array
    # [days, hours, minutes, seconds, time_zone_offset].
    # 'days', 'hours' and 'minutes' are integers.
    # 'seconds' and 'time_zone_offset' are rational numbers.
    # 'days' and 'seconds' are equal to 0 if they're not specified in ((|literal|)).
    # 'time_zone_offset' is equal to nil if not specified.
    #
    # ((|allowDays|)) indicates whether the specification of days is allowed
    # in ((|literal|))
    # ((|allowTimeZone|)) indicates whether the specification of the timeZone is
    # allowed in ((|literal|))
    #
    # All components are returned disregarding the values of ((|allowDays|)) and
    # ((|allowTimeZone|)).
    #
    # Raises an ArgumentError if ((|literal|)) has a bad format.
    def Parser.parse_time_span_and_time_zone(literal, allowDays, allowTimeZone)
      overall_sign = (literal =~ /^-/)? -1 : +1

      if literal =~ /^(([+\-]?\d+)d:)/
        if allowDays
          days = Integer($2)
          days_specified = true
          time_part = literal[($1.length)..-1]
        else
          # detected a day specification in a pure time literal
          raise ArgumentError, "unexpected day specification in #{literal}"
        end
      else
        days = 0;
        days_specified = false
        time_part = literal
      end
			
      # We have to parse the string ourselves because AFAIK :
      #	- strptime() can't parse milliseconds
      #	- strptime() can't parse the time zone custom offset (CET+02:30)
      #	- strptime() accepts trailing chars
      #		(e.g. "12:24-xyz@" ==> "xyz@" is obviously wrong but strptime()
      #		 won't mind)
      if time_part =~ /^([+-]?\d+):(\d+)(?::(\d+)(?:\.(\d+))?)?(?:(?:-([a-zA-Z]+))?(?:([\+\-]\d+)(?::(\d+))?)?)?$/i
        hours = $1.to_i
        minutes = $2.to_i
        # seconds and milliseconds are implemented as one rational number
        # unless there are no milliseconds
        millisecond_part = ($4)? $4.ljust(3, "0") : nil
        if millisecond_part
          seconds = Rational(($3 + millisecond_part).to_i, 10 ** millisecond_part.length)
        else
          seconds = ($3)? Integer($3) : 0
        end
				
        if ($5 or $6) and not allowTimeZone
          raise ArgumentError, "unexpected time zone specification in #{literal}"
        end
				
        time_zone_code = $5 # might be nil

        if $6
          zone_custom_minute_offset = $6.to_i * 60
          if $7
            if zone_custom_minute_offset > 0
              zone_custom_minute_offset = zone_custom_minute_offset + $7.to_i
            else
              zone_custom_minute_offset = zone_custom_minute_offset - $7.to_i
            end
          end
        end

        time_zone_offset = get_time_zone_offset(time_zone_code, zone_custom_minute_offset)
				
        if not allowDays and $1 =~ /^[+-]/
          # unexpected timeSpan syntax
          raise ArgumentError, "unexpected sign on hours : #{literal}"
        end

        # take the sign into account
        hours *= overall_sign if days_specified # otherwise the sign is already applied to the hours
        minutes *= overall_sign
        seconds *= overall_sign
				
        return [ days, hours, minutes, seconds, time_zone_offset ]
				
      else
        raise ArgumentError, "bad time component : #{literal}"
      end
    end
		
    # Parses the given literal (String) into a returned DateTime object.
    #
    # Raises an ArgumentError if ((|literal|)) has a bad format.
    def Parser.parse_date_time(literal)
      raise ArgumentError("date literal is nil") if literal.nil?
			
      begin
        parts = literal.split(" ")
        if parts.length == 1
          return parse_date(literal)
        else
          date = parse_date(parts[0]);
          time_part = parts[1]
					
          days, hours, minutes, seconds, time_zone_offset =
            parse_time_span_and_time_zone(time_part, false, true)
					
          return new_date_time(
            date.year, date.month, date.day, hours, minutes, seconds, time_zone_offset)
        end
				
      rescue ArgumentError
        raise ArgumentError, "Bad date/time #{literal} : #{$!.message}"
      end
    end

    ##
    # Returns the time zone offset (Rational) corresponding to the provided parameters as a fraction
    # of a day. This method adds the two offsets if they are both provided.
    # 
    # +time_zone_code+: can be nil
    # +custom_minute_offset+: can be nil
    #
    def Parser.get_time_zone_offset(time_zone_code, custom_minute_offset)
      return nil unless time_zone_code or custom_minute_offset

      time_zone_offset = custom_minute_offset ? Rational(custom_minute_offset, 60 * 24) : 0
      
      return time_zone_offset unless time_zone_code
			
      # we have to provide some bogus year/month/day in order to parse our time zone code
      d = DateTime.strptime("1999/01/01 #{time_zone_code}", "%Y/%m/%d %Z")
      # the offset is a fraction of a day
      return d.offset() + time_zone_offset
    end
		
    # Parses the +literal+ into a returned Date object.
    #
    # Raises an ArgumentError if +literal+ has a bad format.
    
    def Parser.parse_date(literal)
      # here, we're being stricter than strptime() alone as we forbid trailing chars
      if literal =~ /^(\d+)\/(\d+)\/(\d+)$/
        begin
          return Date.strptime(literal, "%Y/%m/%d")
        rescue ArgumentError
          raise ArgumentError, "Malformed Date <#{literal}> : #{$!.message}"
        end
      end

      raise ArgumentError, "Malformed Date <#{literal}>"
    end
		
    # Returns a String that contains the binary content corresponding to ((|literal|)).
    #
    # ((|literal|)) : a base-64 encoded literal (e.g.
    # "[V2hvIHdhbnRzIHRvIGxpdmUgZm9yZXZlcj8=]")
    def Parser.parse_binary(literal)
      clean_literal = literal[1..-2] # remove square brackets
      return SdlBinary.decode64(clean_literal)
    end
		
    # Parses +literal+ (String) into the corresponding SDLTimeSpan, which is then
    # returned.
    #
    # Raises an ArgumentError if the literal is not a correct timeSpan literal.
    def Parser.parse_time_span(literal)
      days, hours, minutes, seconds, time_zone_offset =
        parse_time_span_and_time_zone(literal, true, false)
			
      milliseconds = ((seconds - seconds.to_i) * 1000).to_i
      seconds = seconds.to_i
			
      return SDLTimeSpan.new(days, hours, minutes, seconds, milliseconds)
    end

    # Close the reader and throw a SdlParseError using the format
    # Was expecting X but got Y.
    #
    def expecting_but_got(expecting, got, line, position)
      @tokenizer.expecting_but_got(expecting, got, line, position)
    end
  end
end
